### PR TITLE
fix(react-refresh): fix incorrect type definitions of `react-refresh`

### PR DESCRIPTION
**Pull request by yuhr**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
	- [`setSignature`](https://github.com/facebook/react/blob/4e8121a75ec7fa838fc781cbab418d40218678a9/packages/react-refresh/src/ReactFreshRuntime.js#L349)
	- [`createSignatureFunctionForTransform`](https://github.com/facebook/react/blob/4e8121a75ec7fa838fc781cbab418d40218678a9/packages/react-refresh/src/ReactFreshRuntime.js#L638)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

File: types/react-refresh/runtime.d.ts [hunk 1]
```diff
@@ -55,6 +55,11 @@ export function hasUnrecoverableErrors(): boolean;
  *    () => [useCustomHook], // Lazy to avoid triggering inline requires
  * );
  */
-export function createSignatureFunctionForTransform(): void;
+export function createSignatureFunctionForTransform(): <T>(
```
Comment (YardenPorat) [inline types/react-refresh/runtime.d.ts:58]: add a test

File: types/react-refresh/runtime.d.ts [hunk 2]
```diff
@@ -55,6 +55,11 @@ export function hasUnrecoverableErrors(): boolean;
...
+    type: T,
+    key: string,
+    forceReset?: boolean,
+    getCustomHooks?: () => AnyFn[],
+) => T | undefined;
```
Comment (YardenPorat) [inline types/react-refresh/runtime.d.ts:63]: maybe add void?

**Review by YardenPorat — CHANGES_REQUESTED** [review]
see comments

File: types/react-refresh/runtime.d.ts (hunk 2 above)
Comment (yuhr) [reply types/react-refresh/runtime.d.ts:63]: I tried `void`, but `pnpm test react-refresh` complained not to use void here.

```
  63:10  error  void is only valid as a return type or generic type argument or the type of a `this` parameter  @typescript-eslint/no-invalid-void-type
```

Comment (codershiba) [reply types/react-refresh/runtime.d.ts:63]: You can disable the rule is `void` is the correct type and needs to be used here.

**Comment by yuhr** [comment]
I've noticed that the [suggested usage in the JSDoc comment](https://github.com/facebook/react/blob/7a2609eedc571049a3272e60d5f7d84601ffca3f/packages/react-refresh/src/ReactFreshRuntime.js#L625) implies that `createSignatureFunctionForTransform` must be also callable without arguments (and the [Babel plugin relies on it](https://github.com/facebook/react/blob/7a2609eedc571049a3272e60d5f7d84601ffca3f/packages/react-refresh/src/ReactFreshBabelPlugin.js#L670)), thus the original type definition in the upstream code is wrong. I extended the signature to allow this usage.

**Comment by yuhr** [comment]
With arguments, [the code path always returns a value](https://github.com/facebook/react/blob/7a2609eedc571049a3272e60d5f7d84601ffca3f/packages/react-refresh/src/ReactFreshRuntime.js#L672), so I removed the unionized `void`.

**Review by YardenPorat — APPROVED** [review]

**Comment by yuhr** [comment]
Ready to merge

**Pull request merged**